### PR TITLE
[Session] Add support for `GetSession` via the CLI + deps

### PR DESCRIPTION
**Pull request by Olshansk**
### Human Summary

- Update the CLI code flow to be able to call `GetSession`
- Add make targets to trigger `GetSession`
- Replace `[]` with `<>` in all CLI functions to differentiate required vs optional params
- Made a couple changes to `SessionHydrator` (and added tests) to account additional error cases

https://github.com/pokt-network/poktroll/assets/1892194/56425906-d06f-4c41-b8a5-d210d8a450cd

### Summary generated by Reviewpad on 08 Nov 23 19:40 UTC
This pull request modifies various files in the codebase. Here is a summary of the changes:

1. The command used for unstaking an application has been updated and the corresponding description has been clarified.
2. The `TestGenesis` function in the `x/supplier/genesis_test.go` file has been modified to replace the `ServiceId` field with the `Service` field in the `SupplierServiceConfig` struct.
3. The `MsgServerStakeApplication` function in the `x/application/keeper/msg_server_stake_application_test.go` file has undergone changes related to the `ServiceConfigs` array, replacing the `ServiceId` field with the `Service` field in multiple instances.
4. The file diff includes changes related to the `x/supplier/genesis_test.go` file, specifically replacing the `ServiceId` field with the `Service` field in the `SupplierServiceConfig` struct and updating the `Id` and `Endpoints` fields.
5. The `application.proto` file has been updated with changes to the comment of the `service_configs` field.
6. The `service.go` file includes changes such as importing a new package, adding new functions, and updating existing functions related to service validation.
7. The `errors.go` file in the `x/supplier/types` package has new error constants related to invalid session start height and ID.
8. The `query_gateway.go` file has changes to the `CmdShowGateway` function, including updates to the `Use`, `Short`, `Args`, and `RunE` fields.
9. The `query_gateway.go` file includes changes to the `CmdShowGateway` function, modifying the `Use`, `Short`, `Args`, and `RunE` fields.
10. The `server_builder.go` file has changes related to the `BuildProvidedServices` function, such as renaming the `serviceId` variable to `service` and creating a new slice called `serviceEndpoints`.
11. The `msg_server_undelegate_from_gateway_test.go` file contains changes to multiple test cases, replacing the `ServiceId` field with the `Service` field.
12. The `services.go` file includes changes such as importing a new package, adding a new function for service validation, and updating existing functions related to service validation.
13. The `go.mod` file has changes related to dependencies, including additions and removals of dependencies.
14. The `tx_submit_proof.go` file includes changes such as reordering import statements, renaming variables, and modifying the `CmdSubmitProof` function.
15. There are changes to the session hydrator functionality, including imports, field replacements, error handling updates, and validation checks.
16. Changes to the `tx_stake_application.go` file involve modifying the usage string and providing instructions for the stake application command.
17. The `service_configs.go` file has had its validation code replaced by validation code for the service instead of the serviceId.
18. The `NewMsgStakeApplication` function in the `message_stake_application.go` file has been modified to change the `ServiceId` field name to `Service` in the `sharedtypes.ApplicationServiceConfig` struct.

Please let me know if you would like more information on any specific part of the diff.

## Issue

- #15

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Testing

- [ ] **Run all unit tests**: `make go_develop_and_test`
- [ ] **Verify Localnet manually**: See the instructions [here](TODO: add link to instructions)

## Sanity Checklist

- [ ] I have tested my changes using the available tooling
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, updated documentation and left TODOs throughout the codebase

### CODE REVIEW

**Review by Olshansk — COMMENTED** [review]
Completed self review

File: Makefile [hunk 1]
```diff
@@ -356,6 +356,26 @@ supplier2_unstake: ## Unstake supplier2
...
+### Session ###
+###############
+
+.PHONY: get_session
+get_session: ## Retrieve the session given the following env vars: (APP_ADDR, SVC, HEIGHT)
+	pocketd --home=$(POCKETD_HOME) q session get-session $(APP) $(SVC) $(HEIGHT) --node $(POCKET_NODE)
+
+.PHONY: get_session_app1_anvil
+get_session_app1_anvil: ## Retrieve the session for (app1, anvil, latest_height)
+	APP=pokt1mrqt5f7qh8uxs27cjm9t7v9e74a9vvdnq5jva4 SVC=anvil HEIGHT=0 make get_session
```
Comment (bryanchriswhite) [inline Makefile:369]: Wdyt about having a make target which looks up the addresses by name with something like `poktrolld keys show -a <name>`?

I don't have a strong opinion but I think it MAY save us some confusion down the road. Could just be a `TODO_CONSIDERATION` comment or something for now.

File: docs/static/openapi.yml
Comment (bryanchriswhite) [inline docs/static/openapi.yml:1]: Why was this deleted? Does it make sense for this to be under version control since it can be generated on demand? Is anyone trying to look at this in the repo?

File: proto/pocket/shared/service.proto [hunk 2]
```diff
@@ -8,6 +8,8 @@ option go_package = "pocket/x/shared/types";
 
 // TODO_CLEANUP(@Olshansk): Add native optional identifiers once its supported; https://github.com/ignite/cli/issues/3698
 
+// TODO_CONSIDERATION: Consider renaming `ServiceId` to `Service` to avoid the confusion of `ServiceId.Id`. This will
```
Comment (bryanchriswhite) [inline proto/pocket/shared/service.proto:11]: :+1:

File: x/application/client/cli/query_application.go [hunk 3]
```diff
@@ -46,7 +46,7 @@ func CmdListApplication() *cobra.Command {
 
 func CmdShowApplication() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:   "show-application [address]",
+		Use:   "show-application <application_address>",
```
Comment (bryanchriswhite) [inline x/application/client/cli/query_application.go:49]: :pray: this has bothered me from the first scaffold. :sweat_smile:

File: x/application/client/cli/tx_unstake_application.go [hunk 4]
```diff
@@ -16,7 +16,7 @@ var _ = strconv.Itoa(0)
 func CmdUnstakeApplication() *cobra.Command {
 	// fromAddress & signature is retrieved via `flags.FlagFrom` in the `clientCtx`
 	cmd := &cobra.Command{
-		Use:   "unstake-application [amount]",
+		Use:   "unstake-application",
```
Comment (bryanchriswhite) [inline x/application/client/cli/tx_unstake_application.go:19]: Nice catch! :sunglasses:

File: x/session/client/cli/query_get_session.go
Comment (bryanchriswhite) [inline x/session/client/cli/query_get_session.go:1]: :fire:

File: docs/static/openapi.yml
Comment (Olshansk) [reply docs/static/openapi.yml:1]: It wasn't deleted. Seems like `ignite generate openapi --yes` hasn't been  executed in a while so `-1354` is just an updated.

```bash
cat docs/static/openapi.yml | wc -l
   78127
```

I'm adding the following make target to simplify things:

```bash
.PHONY: ignite_openapi_gen
ignite_openapi_gen: ## Generate the OpenAPI spec for the Ignite API
	ignite generate openapi --yes
```

File: x/session/keeper/query_get_session.go [hunk 5]
```diff
@@ -15,9 +15,19 @@ func (k Keeper) GetSession(goCtx context.Context, req *types.QueryGetSessionRequ
...
+		return nil, err
+	}
+
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	sessionHydrator := NewSessionHydrator(req.ApplicationAddress, req.ServiceId.Id, req.BlockHeight)
+	// If block height is not specified, use the current (context's latest) block height
+	blockHeight := req.BlockHeight
+	if blockHeight == 0 {
+		blockHeight = ctx.BlockHeight()
```
[the remark this answers is not in the document]
Comment (Olshansk) [reply x/session/keeper/query_get_session.go:27]: > he height of the block which is currently being validated, as opposed to the height of the last committed block.

This is part of the `QueryClient`, not the `MsgServer`.

A `Msg` (i.e. a Tx creating a state transition) requires validation.

A `Query` (i.e. just a read to the on-chain) depends on how far up the node has synched.

Claim & Proof lifecycle functions are `Msg` that lead to state transition.

`GetSession` is a query that does not lead to a state transition.

Does that make sense?

File: Makefile [hunk 6]
```diff
@@ -410,3 +441,14 @@ go_docs: check_godoc ## Generate documentation for the project
 	echo "Visit http://localhost:6060/pkg/pocket/"
 	godoc -http=:6060
 
+.PHONY: ignite_openapi_gen
+ignite_openapi_gen: ## Generate the OpenAPI spec for the Ignite API
```
Comment (bryanchriswhite) [inline Makefile:445]: Wdyt about just `openapi_gen` as this is (rightly so IMHO) under the "documentation" section, as opposed to the "ignite helpers" section? :thinking:

File: Makefile [hunk 7]
```diff
@@ -384,11 +413,13 @@ acc_balance_query_module_app: ## Query the balance of the network level "applica
 
 .PHONY: acc_balance_query_module_supplier
 acc_balance_query_module_supplier: ## Query the balance of the network level "supplier" module
-	make acc_balance_query ACC=pokt1j40dzzmn6cn9kxku7a5tjnud6hv37vesr5ccaa
+	SUPPLIER1=$(make poktrolld_addr ACC_NAME=supplier1)
+	make acc_balance_query ACC=SUPPLIER1
```
Comment (bryanchriswhite) [inline Makefile:417]: :rocket: :pray:

File: Makefile [hunk 8]
```diff
@@ -410,3 +441,14 @@ go_docs: check_godoc ## Generate documentation for the project
...
+.PHONY: ignite_openapi_gen
+ignite_openapi_gen: ## Generate the OpenAPI spec for the Ignite API
+	ignite generate openapi --yes
+
+######################
+### Ignite Helpers ###
+######################
+
+.PHONY: poktrolld_addr
+poktrolld_addr: ## Retrieve the address for an account by ACC_NAME
```
Comment (bryanchriswhite) [inline Makefile:453]: :rocket::rocket::rocket:

File: x/gateway/client/cli/tx_stake_gateway.go [hunk 9]
```diff
@@ -3,20 +3,20 @@ package cli
...
+
+	"github.com/pokt-network/poktroll/x/gateway/types"
 )
 
 var _ = strconv.Itoa(0)
 
 func CmdStakeGateway() *cobra.Command {
 	cmd := &cobra.Command{
-		Use:   "stake-gateway [amount]",
+		Use:   "stake-gateway <upokt_amount>",
```
Comment (bryanchriswhite) [inline x/gateway/client/cli/tx_stake_gateway.go:19]: :ok_hand:

File: x/session/keeper/query_get_session_test.go [hunk 10]
```diff
@@ -75,6 +76,7 @@ func TestSession_GetSession_Success(t *testing.T) {
 
 func TestSession_GetSession_Failure(t *testing.T) {
 	keeper, ctx := keepertest.SessionKeeper(t)
+	ctx = ctx.WithBlockHeight(100) // provide a sufficiently large block height to avoid errors
```
Comment (bryanchriswhite) [inline x/session/keeper/query_get_session_test.go:79]: What does this actually do?

File: x/session/keeper/query_get_session_test.go [hunk 11]
```diff
@@ -91,34 +93,69 @@ func TestSession_GetSession_Failure(t *testing.T) {
...
+			expectedErrContains: types.ErrSessionInvalidAppAddress.Error(),
+		},
+		{
+			name: "service ID is invalid",
+
+			appAddr:     keepertest.TestApp1Address,
+			serviceId:   "invalid_serviceId",
+			blockHeight: 1,
+
+			expectedErrContains: "invalid serviceID for session being retrieved",
```
Comment (bryanchriswhite) [inline x/session/keeper/query_get_session_test.go:136]: Not sure if it's relevant and/or there's any action item  but I just  wanted to call out that I think this might also be invalid because it's too long. IIRC, I saw somewhere that the max length for a serviceId was 7 or 8.

File: Makefile (hunk 7 above)
Comment (Olshansk) [reply Makefile:417]: Thank you fo the suggestion!

File: x/session/keeper/query_get_session.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply x/session/keeper/query_get_session.go:27]: I #PUCed it up

File: x/session/keeper/query_get_session_test.go (hunk 10 above)
Comment (Olshansk) [reply x/session/keeper/query_get_session_test.go:79]: It literally just sets a variable to a value for when we test things.

```go
// WithBlockHeight returns a Context with an updated block height.
func (c Context) WithBlockHeight(height int64) Context {
	newHeader := c.BlockHeader()
	newHeader.Height = height
	return c.WithBlockHeader(newHeader)
}
```

File: x/session/keeper/query_get_session_test.go (hunk 11 above)
Comment (Olshansk) [reply x/session/keeper/query_get_session_test.go:136]: Updated the serviceId to `serviceId:   "service_id_is_too_long_to_be_valid",`

**Comment by Olshansk** [comment]
@bryanchriswhite PTAL again when you have a chance. This PR should be really close to ready!

File: Makefile [hunk 12]
```diff
@@ -311,31 +311,37 @@ app_delegate: ## Delegate trust to a gateway (must specify the APP and GATEWAY_A
 
 .PHONY: app1_delegate_gateway1
 app1_delegate_gateway1: ## Delegate trust to gateway1
-	APP=app1 GATEWAY_ADDR=pokt15vzxjqklzjtlz7lahe8z2dfe9nm5vxwwmscne4 make app_delegate
+	GATEWAY1=$(make poktrolld_addr ACC_NAME=gateway1)
+	APP=app1 GATEWAY_ADDR=$(GATEWAY1) make app_delegate
```
Comment (bryanchriswhite) [inline Makefile:315]: Does this  need the double dollar-sign as well?:
```suggestion
	GATEWAY1=$$(make poktrolld_addr ACC_NAME=gateway1)
	APP=app1 GATEWAY_ADDR=$$(GATEWAY1) make app_delegate
```

Comment (bryanchriswhite) [inline Makefile:315]: 1. While separating the variable assignments with `&&` seems to be optional, applying the `\` to continue the line seems not to be.
2. This target missed the bash var --> make var refactor pass.
3. This pattern can be simplified to [2]

[1]
```suggestion
	GATEWAY1=$$(make poktrolld_addr ACC_NAME=gateway1) \
	APP=app1 GATEWAY_ADDR=$$GATEWAY1 make app_delegate
```
[2]
```suggestion
	GATEWAY_ADDR=$$(make poktrolld_addr ACC_NAME=gateway1) \
	APP=app1 make app_delegate
```

File: Makefile [hunk 13]
```diff
@@ -311,31 +311,37 @@ app_delegate: ## Delegate trust to a gateway (must specify the APP and GATEWAY_A
...
 
 .PHONY: app2_delegate_gateway2
 app2_delegate_gateway2: ## Delegate trust to gateway2
-	APP=app2 GATEWAY_ADDR=pokt15w3fhfyc0lttv7r585e2ncpf6t2kl9uh8rsnyz make app_delegate
+	GATEWAY2=$$(make poktrolld_addr ACC_NAME=gateway2) && \
```
Comment (bryanchriswhite) [inline Makefile:319]: `&&` seems to be unnecessary from my testing.

File: Makefile [hunk 14]
```diff
@@ -311,31 +311,37 @@ app_delegate: ## Delegate trust to a gateway (must specify the APP and GATEWAY_A
...
+	APP=app2 GATEWAY_ADDR=$$GATEWAY2 make app_delegate
```
Comment (bryanchriswhite) [inline Makefile:320]: Same as 2. above :point_up_2: - any reason not to simplify this?

File: x/session/client/cli/query_get_session_test.go [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+
+	require.Len(t, supplierSvc0.Services, 1)
+	require.Len(t, supplierSvc1.Services, 1)
+
+	require.Equal(t, supplierSvc0.Services[0].ServiceId.Id, "svc0")
+	require.Equal(t, supplierSvc1.Services[0].ServiceId.Id, "svc1")
+
+	// Prepare the test cases
+	tests := []struct {
+		desc string
```
Comment (bryanchriswhite) [inline x/session/client/cli/query_get_session_test.go:49]: :eyes:

File: x/session/keeper/session_hydrator_test.go [hunk 16]
```diff
@@ -86,20 +91,33 @@ func TestSession_HydrateSession_Metadata(t *testing.T) {
...
 	sessionKeeper, ctx := keepertest.SessionKeeper(t)
+	ctx = ctx.WithBlockHeight(100) // provide a sufficiently large block height to avoid errors
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			sessionHydrator := keeper.NewSessionHydrator(appAddr, serviceId, tt.blockHeight)
 			session, err := sessionKeeper.HydrateSession(ctx, sessionHydrator)
+
+			if tt.errExpected {
+				require.Error(t, err)
```
Comment (bryanchriswhite) [inline x/session/keeper/session_hydrator_test.go:118]: Wdyt  about making `tt.errExpected` an `error` type and doing something like:
```suggestion
				require.ErrorIs(t, tt.errExpected, err)
```

File: x/session/keeper/session_hydrator_test.go [hunk 17]
```diff
@@ -198,28 +217,39 @@ func TestSession_HydrateSession_SessionId(t *testing.T) {
 // TODO_TECHDEBT: Expand these tests to account for application joining/leaving the network at different heights as well changing the services they support
 func TestSession_HydrateSession_Application(t *testing.T) {
 	type test struct {
-		name    string
-		appAddr string
+		name      string
```
Comment (bryanchriswhite) [inline x/session/keeper/session_hydrator_test.go:220]: :eyes:

File: x/session/types/errors.go [hunk 18]
```diff
@@ -8,7 +8,11 @@ import (
 
 // x/session module sentinel errors
 var (
-	ErrHydratingSession  = sdkerrors.Register(ModuleName, 1, "error during session hydration")
-	ErrAppNotFound       = sdkerrors.Register(ModuleName, 2, "application not found")
-	ErrSuppliersNotFound = sdkerrors.Register(ModuleName, 3, "suppliers not found")
+	ErrHydratingSession          = sdkerrors.Register(ModuleName, 1, "error during session hydration")
+	ErrAppNotFound               = sdkerrors.Register(ModuleName, 2, "application not found")
+	ErrAppNotStakedForService    = sdkerrors.Register(ModuleName, 3, "application not staked for service")
+	ErrSuppliersNotFound         = sdkerrors.Register(ModuleName, 4, "suppliers not found")
```
[the remark this answers is not in the document]
Comment (bryanchriswhite) [reply x/session/types/errors.go:14]: I find myself wondering how we should be thinking about which modules should own which errors given how we intend to use them.

File: x/shared/helpers/service_test.go [hunk 19]
```diff
@@ -28,6 +32,78 @@ func TestIsValidServiceId(t *testing.T) {
...
+			expected: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := IsValidService(&tt.input)
+			if got != tt.expected {
+				t.Errorf("IsValidService(%v) = %v, want %v", tt.input, got, tt.expected)
+			}
```
Comment (bryanchriswhite) [inline x/shared/helpers/service_test.go:102]: Any reason not to use `require.Errorf()`?

```suggestion
			require.Equalf(t, tt.expected, got, "IsValidService(%v)",  tt.input)
```

File: x/shared/helpers/service_test.go [hunk 20]
```diff
@@ -28,6 +32,78 @@ func TestIsValidServiceId(t *testing.T) {
...
+		t.Run(test.input, func(t *testing.T) {
+			result := IsValidServiceName(test.input)
+			if result != test.expected {
+				t.Errorf("For input %s, expected %v but got %v", test.input, test.expected, result)
+			}
+		})
+	}
+}
+
+func TestIsValidService(t *testing.T) {
```
Comment (bryanchriswhite) [inline x/shared/helpers/service_test.go:59]: I think this is [duplicated in #150](https://github.com/pokt-network/poktroll/pull/150/files/2b95bc48e7eb8651701937a2fbf4231762e8e96b#diff-787693a5c2759a1d7cfa40155745aadc2ff109cb49bfe446952aea388adeea67R9).

File: x/shared/helpers/service_test.go [hunk 21]
```diff
@@ -28,6 +32,78 @@ func TestIsValidServiceId(t *testing.T) {
 	}
 }
 
+func TestIsValidServiceName(t *testing.T) {
+	tests := []struct {
+		input    string
+		expected bool
+	}{
```
Comment (bryanchriswhite) [inline x/shared/helpers/service_test.go:39]: I think this would read nicer if we cleaned it up like the rest of the table-driven tests.

File: x/session/keeper/query_get_session.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (h5law) [reply x/session/keeper/query_get_session.go:32]: Top tier comment

**Review by h5law — CHANGES_REQUESTED** [review]
Overall great PR, only things standing out were some of the comments @bryanchriswhite had already left. Once they are in this is good to go! :shipit:

File: x/shared/helpers/service_test.go (hunk 20 above)
Comment (Olshansk) [reply x/shared/helpers/service_test.go:59]: Good point. Let's merge #150 in first and then I'll resolve the conflicts.

File: x/shared/helpers/service_test.go (hunk 21 above)
Comment (Olshansk) [reply x/shared/helpers/service_test.go:39]: Deferring to #150, and will rebase on top of that

File: x/session/types/errors.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (Olshansk) [reply x/session/types/errors.go]: Addressed this for consistency. Lmk what you think

File: x/session/keeper/session_hydrator_test.go [hunk 22]
```diff
@@ -25,81 +26,98 @@ func TestSession_HydrateSession_Success_BaseCase(t *testing.T) {
...
 			blockHeight: 5,
 
 			expectedNumBlocksPerSession: 4,
 			expectedSessionNumber:       1,
 			expectedSessionStartBlock:   4,
+			errExpected:                 nil,
+		},
+		{
+			desc:        "blockHeight > contextHeight",
+			blockHeight: 9001, // block height over 9000 is too height given that the context height is 100
```
Comment (bryanchriswhite) [inline x/session/keeper/session_hydrator_test.go:101]: wat 9000?

File: x/session/types/query_get_session_request.go [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+	sdkerrors "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	sharedhelpers "github.com/pokt-network/poktroll/x/shared/helpers"
+	sharedtypes "github.com/pokt-network/poktroll/x/shared/types"
+)
+
+// NOTE: Please note that `QueryGetSessionRequest` is not a `sdk.Msg`, and is therefore not a message/request
+// that will be signable or invoke a state transition. However, following a similar `ValidateBasic` pattern
+// allows us to localize & reuse validation logic.
```
Comment (bryanchriswhite) [inline x/session/types/query_get_session_request.go:13]: :100: Nice comment & design.

**Review by bryanchriswhite — APPROVED** [review]
LFG

Let's gooooooo! :rocket:

File: x/session/types/query_get_session_request.go (hunk 23 above)
Comment (Olshansk) [reply x/session/types/query_get_session_request.go:13]: Ty!

File: x/session/keeper/session_hydrator_test.go (hunk 22 above)
Comment (Olshansk) [reply x/session/keeper/session_hydrator_test.go:101]: https://twitter.com/olshansky/status/1722339951466217476

**Comment by Olshansk** [comment]
I accidently merged this into the wrong branch so attempt #2 it in #163.
